core/gateway.go

```diff
@@ -28,23 +30,55 @@ const (
 // of execs when they fail.
 type GatewayClient struct {
 	bkgw.Client
+	refs map[*ref]struct{}
+	mu   sync.Mutex
+}
+
+func NewGatewayClient(baseClient bkgw.Client) *GatewayClient {
+	return &GatewayClient{
+		Client: baseClient,
+		refs:   make(map[*ref]struct{}),
+	}
 }
 
 func (g *GatewayClient) Solve(ctx context.Context, req bkgw.SolveRequest) (_ *bkgw.Result, rerr error) {
 	defer wrapSolveError(&rerr, g.Client)
+	req.CacheImports = []bkgw.CacheOptionsEntry{{
+		Type: "dagger",
+	}}
 	res, err := g.Client.Solve(ctx, req)
 	if err != nil {
 		return nil, err
 	}
+	g.mu.Lock()
+	defer g.mu.Unlock()
 	if res.Ref != nil {
-		res.Ref = &ref{Reference: res.Ref, gw: g}
+		r := &ref{Reference: res.Ref, gw: g}
+		g.refs[r] = struct{}{}
+		res.Ref = r
 	}
 	for k, r := range res.Refs {
-		res.Refs[k] = &ref{Reference: r, gw: g}
+		r := &ref{Reference: r, gw: g}
+		g.refs[r] = struct{}{}
+		res.Refs[k] = r
 	}
 	return res, nil
 }
 
+// CombinedResult returns a buildkit result with all the refs solved by this client so far.
+// This is useful for constructing a result for remote caching.
+func (g *GatewayClient) CombinedResult() *bkgw.Result {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	res := bkgw.NewResult()
+	i := 0
+	for r := range g.refs {
+		res.AddRef(strconv.Itoa(i), r.Reference)
```

Comment:
Is it okay if it's non-deterministic? (e.g. same pipeline might yield a different reference set)

Reply:
To my knowledge yes, though it's hard to trace through every usage in the buildkit codebase so I am not 100% sure yet.

I guess we could sort on a hash of the LLB if needed in the future.